api/processors/actions/student/left_project_prematurely.py

```diff
@@ -0,0 +1,34 @@
+from django.db import transaction
+
+from api.models import Student
+from api.models.choices.log_event_type import StudentLogEventType
+from api.models.choices.status import StudentProjectStatus
+from api.models.log_event import StudentLogEvent
+from api.processors.actions.student import StudentActionProcessor
+
+
+class StudentLeftProjectPrematurelyProcessor(StudentActionProcessor):
+    def __init__(self, student: Student):
+        super().__init__(student)
+
+    @transaction.atomic
+    def process(self) -> None:
+        self._create_log_events()
+        self._update_groups()
+        self._set_statuses()
+
+    def _update_groups(self) -> None:
+        for group in self.student.groups.all():
+            group.students.remove(self.student)
+            group.students_former.add(self.student)
```

Comment:
I think you can actually simply remove every group from `student.groups` and add it to `student,groups_former`. Right now we're sort of going back and forth too many times in this relationship.